go/database/mpt/proof.go

```diff
@@ -0,0 +1,197 @@
+// encoded as an RLP byte slice.
+type ProofElement struct {
+	hash common.Hash
+	node RlpNode
+}
+
+// Equal returns true if the input proof element is equal to the current proof element.
+func (p ProofElement) Equal(other ProofElement) bool {
+	return p.hash == other.hash && bytes.Equal(p.node, other.node)
+}
+
+// Hash returns the hash of the proof element.
+func (p ProofElement) Hash() common.Hash {
+	return p.hash
+}
+
+// Rlp returns the RLP encoded node of the proof element.
+func (p ProofElement) Rlp() RlpNode {
+	return p.node
+}
+
+// proof is a chain of MPT nodes and their hashes that represent a witness proof.
+// It is used to verify the proof by checking that the nodes of the MPT are present in this proof.
+// The proof is created by iterating the MPT from the root down to the leaf node following
+// the address or a storage key.
+// Each node encountered on the path is hashed and stored in the proof.
+type proof []ProofElement
+
+// nodeVisitor is a visitor that visits MPT nodes and creates a witness proof.
+// It uses the hasher and node source to hash and encode the nodes.
+// It computes RLP and hash of nodes that compose the proof.
+type nodeVisitor struct {
+	proof      proof
+	hasher     hasher
+	nodeSource NodeSource
+	err        error
+}
+
+// Visit computes RLP and hash of the visited node and appends it to the proof.
+// It is expected that the nodes are provided via the visitor in the same order
+// as they are present in the MPT path from root to the leaf to prove.
+func (p *nodeVisitor) Visit(node Node, _ NodeInfo) VisitResponse {
+	data := make([]byte, 0, 1024)
+	rlp, err := p.hasher.encode(node, p.nodeSource, data)
+	if err != nil {
+		p.err = err
+		return VisitResponseAbort
+	}
+	hash := common.Keccak256(rlp)
+
+	p.proof = append(p.proof, ProofElement{hash, rlp})
+
+	return VisitResponseContinue
+}
+
+// StorageProof represents a witness proof for a storage slot.
+type StorageProof proof
+
+// AccountProof represents a witness proof for an account.
+type AccountProof proof
+
+// ProofSource is a utility to create witness proofs for accounts and storage slots.
+// It has access to nodes of the MPT and a root.
+// The witness proof is created by iterating the MPT from the current root
+// down to a leaf node following the input address or  a storage key.
+// Each node encountered on the path is hashed and stored in a proof.
+// This proof represents the witness proof of the input address and storage key.
+// This proof is furthermore verified by iterating it as an MPT structure from the root down
+// to the leaf node while checking at every step that the encountered node is present in
+// the proof.
+type ProofSource struct {
+	nodeSource NodeSource
+	hasher     hasher
+	root       *NodeReference
+}
+
+// NewProofSource creates a new proof source with the input node source and root node.
+func NewProofSource(root *NodeReference, nodeSource NodeSource) *ProofSource {
```

Comment:
Should the `ProofSource` be made an implementation detail and the function
```
func CreateStorageProofs(
        source NodeSource,
        root *NodeReference,
        address common.Address, 
        keys ...common.Key,
) (StorageProofs, error) { ... }
```
become a free function? Same for `CreateAccountProof`?

I do not think there is a utility in having the ProofSource as an independent entity. Wdyt?

Reply:
done